src/utils.ts

```diff
@@ -0,0 +1,564 @@
+      success: false,
+      message: 'Could not validate our best block',
+      error: ourBestBlockInFullNode.message,
+    };
+
+    return;
+  }
+
+  const { meta } = ourBestBlockInFullNode;
+
+  if ((meta.voided_by &&
+       meta.voided_by.length &&
+         meta.voided_by.length > 0)) {
+    yield {
+      type: 'reorg',
+      success: false,
+      message: 'Our best block was voided, we should reorg.',
+    };
+
+    return;
+  }
+
+  if (ourBestBlock.height > meta.height) {
+    yield {
+      type: 'reorg',
+      success: false,
+      message: 'Our height is higher than the wallet-service\'s height, we should reorg.',
+    };
+
+    return;
+  }
+
+  console.log('Best block is valid.');
+  let success = true;
+
+  blockLoop:
+    for (let i = ourBestBlock.height + 1; i <= fullNodeBestBlock.height; i++) {
+    const block: FullBlock = await downloadBlockByHeight(i);
+    const preparedBlock: PreparedTx = prepareTx(block);
+
+    // Ignore parents[0] because it is a block
+    const blockTxs = [
+      block.parents[1],
+      block.parents[2],
+    ];
+
+    // Download block transactions
+    const txs: FullTx[] = await recursivelyDownloadTx(block.txId, blockTxs);
+
+    // We will send the block only after all transactions were downloaded
+    // to be sure that all downloads were succesfull since there is no
+    // ROLLBACK yet on the wallet-service.
+    const sendBlockResponse: ApiResponse = await sendTx(preparedBlock);
+
+    if (!sendBlockResponse.success) {
+      console.log(sendBlockResponse);
+      yield {
+        type: 'error',
+        success: false,
+        message: `Failure on block ${preparedBlock.tx_id}`,
+      };
+
+      success = false;
+
+      break;
+    }
+
+    // Exclude duplicates:
+    const uniqueTxs: FullTx[] = txs.reduce((acc: FullTx[], tx: FullTx): FullTx[] => {
+      const alreadyInAcc = acc.find((accTx) => accTx.txId === tx.txId);
+
+      if (alreadyInAcc) return acc;
+
+      return [...acc, tx];
+    }, []);
+
+    txLoop:
+      for (let i = 0; i < uniqueTxs.length; i++) {
+      const preparedTx: PreparedTx = prepareTx(uniqueTxs[i]);
+
+      try {
+        const sendTxResponse: ApiResponse = await sendTx(preparedTx);
+
+        if (!sendTxResponse.success) {
+          throw new Error(sendTxResponse.message);
+        }
+      } catch (e) {
+        yield {
+          type: 'transaction_failure',
+          success: false,
+          message: `Failure on transaction ${preparedTx.tx_id} from block: ${preparedBlock.tx_id}`,
+        };
+
+        success = false;
+
+        break blockLoop;
+      }
+    }
+
+    yield {
+      type: 'block_success',
+      success: true,
+      blockId: preparedBlock.tx_id,
+      height: preparedBlock.height,
+      transactions: uniqueTxs.map((tx: FullTx) => {
+        return tx.txId;
+      }),
+    };
+  }
+
+  yield {
+    success,
+    type: 'finished',
+  };
+}
+
+export class LRU {
+  max: number;
+  cache: Map<string, any>;
```

Comment:
Just leave a note that `Map` remembers the original insertion order.

Reply:
Done in 1b3435e83e095882bd247ff524c68a6fbc2c6612